File: frontend/src/services/attendanceService.ts
import api from './api';
import type { Attendance, AttendanceSummary, AttendancePolicy } from '../types';

export const attendanceService = {
  getTodayStatus: () =>
    api.get<Attendance>('/attendance/today').then((r) => r.data),

  clockIn: () =>
    api.post<Attendance>('/attendance/clock-in').then((r) => r.data),

  clockOut: () =>
    api.post<Attendance>('/attendance/clock-out').then((r) => r.data),

  getPersonalHistory: (startDate: string, endDate: string) =>
    api.get<Attendance[]>('/attendance/history', { params: { startDate, endDate } }).then((r) => r.data),

  getPersonalSummary: () =>
    api.get<AttendanceSummary>('/attendance/summary').then((r) => r.data),

  getTeamTodayStatus: () =>
    api.get<Attendance[]>('/attendance/team/today').then((r) => r.data),

  getTeamHistory: (startDate: string, endDate: string) =>
    api.get<Attendance[]>('/attendance/team/history', { params: { startDate, endDate } }).then((r) => r.data),

  getTeamSummaryCounts: () =>
    api.get<Record<string, number>>('/attendance/team/summary').then((r) => r.data),

  getPolicy: () =>
    api.get<AttendancePolicy>('/attendance-policy').then((r) => r.data),

  updatePolicy: (data: AttendancePolicy) =>
    api.put<AttendancePolicy>('/attendance-policy', data).then((r) => r.data),
};

File: src/main/java/com/ems/employee_management_system/service/AttendanceService.java
package com.ems.employee_management_system.service;

import com.ems.employee_management_system.dto.AttendanceDTO;
import com.ems.employee_management_system.dto.AttendanceSummaryDTO;
import java.time.LocalDate;
import java.util.List;
import java.util.Map;

/**
 * Service interface for employee punch operations and attendance statistics.
 */
public interface AttendanceService {
    AttendanceDTO clockIn(String username);
    AttendanceDTO clockOut(String username);
    AttendanceDTO getTodayStatus(String username);
    List<AttendanceDTO> getPersonalHistory(String username, LocalDate startDate, LocalDate endDate);
    AttendanceSummaryDTO getPersonalSummary(String username);
    List<AttendanceDTO> getTeamTodayStatus(String managerUsername);
    List<AttendanceDTO> getTeamHistory(String managerUsername, LocalDate startDate, LocalDate endDate);
    Map<String, Long> getTeamSummaryCounts(String managerUsername);
}
